linode_api4/objects/database.py

```diff
@@ -0,0 +1,405 @@
+from linode_api4.objects import Base, Property, MappedObject, DerivedBase, UnexpectedResponseError
+
+
+class DatabaseType(Base):
+    api_endpoint = '/databases/types/{id}'
+
+    properties = {
+        'deprecated': Property(filterable=True),
+        'disk': Property(),
+        'engines': Property(),
+        'id': Property(identifier=True),
+        'label': Property(),
+        'memory': Property(),
+        'vcpus': Property(),
+    }
+
+    def _populate(self, json):
+        """
+        Allows changing the name "class" in JSON to "type_class" in python
+        """
+        super()._populate(json)
+
+        if 'class' in json:
+            setattr(self, 'type_class', json['class'])
+        else:
+            setattr(self, 'type_class', None)
+
+
+class DatabaseEngine(Base):
+    api_endpoint = '/databases/engines/{id}'
+
+    properties = {
+        'id': Property(identifier=True),
+        'engine': Property(filterable=True),
+        'version': Property(filterable=True),
+    }
+
+
+class Database(Base):
+    """
+    A generic Database instance.
+    """
+
+    api_endpoint = '/databases/instances/{id}'
+
+    properties = {
+        'id': Property(),
+        'label': Property(),
+        'allow_list': Property(),
+        'cluster_size': Property(),
+        'created': Property(),
+        'encrypted': Property(),
+        'engine': Property(),
+        'hosts': Property(),
+        'instance_uri': Property(),
+        'region': Property(),
+        'status': Property(),
+        'type': Property(),
+        'updated': Property(),
+        'updates': Property(),
+        'version': Property(),
+    }
+
+    @property
+    def instance(self):
+        """
+        Returns the underlying database object for the corresponding database
+        engine. This is useful for performing operations on generic databases.
+
+        The following is an example of printing credentials for all databases regardless of engine::
+
+            client = LinodeClient(TOKEN)
+
+            databases = client.database.instances()
+
+            for db in databases:
+                print(f"{db.hosts.primary}: {db.instance.credentials.username} {db.instance.credentials.password}")
+        """
+
+        if not hasattr(self, '_instance'):
+            engine_type_translation = {
+                'mysql': MySQLDatabase,
+                'postgresql': PostgreSQLDatabase,
+            }
+
+            if self.engine not in engine_type_translation:
+                return None
+
+            self._set('_instance', engine_type_translation[self.engine](self._client, self.id))
+
+        return self._instance
+
+    def invalidate(self):
+        """
+        Clear out cached properties.
+        """
+
+        for attr in ['_instance']:
+            if hasattr(self, attr):
+                delattr(self, attr)
+
+        Base.invalidate(self)
+
+
+class DatabaseBackup(DerivedBase):
```

Comment:
You should include a comment noting that this is not intended to be used on its own